File: java/src/main/java/it/giannibombelli/wsc2026/payment/integration/PaymentResultIntegrationEvent.java
package it.giannibombelli.wsc2026.payment.integration;

import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.utils.Require;

/**
 * Published Language esposta dal BC {@code payment} verso i BC downstream.
 * <p>
 * Rappresenta il risultato di un pagamento senza esporre gli eventi di dominio interni di {@code payment}.
 * I campi usano solo tipi stabili ({@code String}, {@link Money}) in modo da essere consumabili
 * da qualsiasi bounded context senza introdurre coupling sui tipi interni di {@code payment}.
 */
public sealed interface PaymentResultIntegrationEvent {

    record PaymentAcceptedIntegrationEvent(
        String clientReference,
        Money amount
    ) implements PaymentResultIntegrationEvent {
        public PaymentAcceptedIntegrationEvent {
            Require.requireArgument(clientReference, "clientReference");
            Require.requireArgument(amount, "amount");
        }
    }

    record PaymentRejectedIntegrationEvent(
        String clientReference,
        Money amount,
        String reason
    ) implements PaymentResultIntegrationEvent {
        public PaymentRejectedIntegrationEvent {
            Require.requireArgument(clientReference, "clientReference");
            Require.requireArgument(amount, "amount");
            Require.requireArgument(reason, "reason");
        }
    }

    record PaymentExpiredIntegrationEvent(
        String clientReference,
        Money amount
    ) implements PaymentResultIntegrationEvent {
        public PaymentExpiredIntegrationEvent {
            Require.requireArgument(clientReference, "clientReference");
            Require.requireArgument(amount, "amount");
        }
    }
}

File: typescript/src/payment/integration/paymentResultIntegrationEvent.ts
// Published Language esposta dal BC payment verso i BC downstream.
// Rappresenta il risultato di un pagamento senza esporre gli eventi di dominio interni di payment.
// I campi usano solo tipi stabili (string, Money) in modo da essere consumabili
// da qualsiasi bounded context senza introdurre coupling sui tipi interni di payment.

import type { Money } from '@/common/domain/primitive/money.js';

export type PaymentAcceptedIntegrationEvent = {
  readonly kind: 'PaymentAcceptedIntegrationEvent';
  readonly clientReference: string;
  readonly amount: Money;
};

export type PaymentRejectedIntegrationEvent = {
  readonly kind: 'PaymentRejectedIntegrationEvent';
  readonly clientReference: string;
  readonly amount: Money;
  readonly reason: string;
};

export type PaymentExpiredIntegrationEvent = {
  readonly kind: 'PaymentExpiredIntegrationEvent';
  readonly clientReference: string;
  readonly amount: Money;
};

export type PaymentResultIntegrationEvent =
  | PaymentAcceptedIntegrationEvent
  | PaymentRejectedIntegrationEvent
  | PaymentExpiredIntegrationEvent;

export function paymentAcceptedIntegrationEvent(
  clientReference: string,
  amount: Money
): PaymentAcceptedIntegrationEvent {
  return { kind: 'PaymentAcceptedIntegrationEvent', clientReference, amount };
}

export function paymentRejectedIntegrationEvent(
  clientReference: string,
  amount: Money,
  reason: string
): PaymentRejectedIntegrationEvent {
  return { kind: 'PaymentRejectedIntegrationEvent', clientReference, amount, reason };
}

export function paymentExpiredIntegrationEvent(
  clientReference: string,
  amount: Money
): PaymentExpiredIntegrationEvent {
  return { kind: 'PaymentExpiredIntegrationEvent', clientReference, amount };
}
